build/source/engine/read_attrb.f90

```diff
@@ -29,304 +29,328 @@ module read_attrb_module
  ! ************************************************************************************************
  ! public subroutine read_dimension: read HRU and GRU dimension information on local attributes
  ! ************************************************************************************************
- subroutine read_dimension(maxGRU,maxHRU,err,message)
+ subroutine read_dimension(attrFile,fileGRU,fileHRU,nGRU,nHRU,err,message,startGRU,checkHRU)
```

Comment:
There are two things to notice here:

1. There is no way to separate the index assignment form the allocate without reading the file twice. Both allocating and assigning requires knowing the HRU/GRU mapping from the Attributes file, and if we do not store that information when we read it in for the allocation then we simply have to read it in again.

2. There is no generality to separating these allocation and assignment functions - these routines are both doing the same task, and cannot be used for any other task. The local and global allocate routines have some generality (although they could be made truly general by passing nSnow, nSoil, nStats, etc. rather than grabbing those from global variables). 

The bottom line is that there is nothing gained in terms of (a) functionality, (b) code readability or parsimony, or (c) run-time efficiency by repeating a file read.  Doing this reduces run-time efficiency, decreases code readability, and if we ever change the input structure, we have to change two redundant sections of code.